File: packages/sb-uhq/src/self/commands/utilitaires/msgbookmarks.js
"use strict";

const fs   = require("fs");
const path = require("path");
const { dataPath } = require("../../func/data-path");

const MSGBM_FILE = dataPath("config", "msgbookmarks.json");

// ── I/O ──────────────────────────────────────────────────────────────────────

function load() {
  try { return JSON.parse(fs.readFileSync(MSGBM_FILE, "utf-8")); }
  catch { return []; }
}

function save(data) {
  fs.mkdirSync(path.dirname(MSGBM_FILE), { recursive: true });
  fs.writeFileSync(MSGBM_FILE, JSON.stringify(data, null, 2));
}

// ── Logique pure ─────────────────────────────────────────────────────────────

/**
 * @param {import("discord.js-selfbot-v13").Client} client
 * @param {{ action: "list"|"add"|"remove"|"note"|"clear", url?: string, note?: string, index?: number }} payload
 */
async function execute(client, payload) {
  const { action, index, note } = payload;
  let bookmarks = load();

  if (action === "list") return { bookmarks };

  if (action === "add") {
    if (!payload.url) throw new Error("url requis.");

    // Parser l'URL Discord : https://discord.com/channels/guildId/channelId/messageId
    const match = payload.url.match(/discord\.com\/channels\/(\d+|@me)\/(\d+)\/(\d+)/);
    if (!match) throw new Error("URL de message Discord invalide.");

    const [, guildId, channelId, messageId] = match;

    // Tenter de récupérer le message pour avoir auteur + contenu
    let authorTag = "Inconnu";
    let content   = "";

    try {
      // Récupérer le salon depuis le cache d'abord, puis fetch si absent
      let channel = client.channels.cache.get(channelId);
      if (!channel) {
        channel = await client.channels.fetch(channelId).catch(() => null);
      }

      if (channel) {
        // Fetch le message directement par son ID
        const msg = await channel.messages.fetch(messageId).catch(() => null);
        if (msg) {
          // Récupérer le tag de l'auteur — username#discriminator ou juste username
          const author = msg.author;
          if (author) {
            // discord.js-selfbot-v13 : discriminator "0" = nouveau système sans #tag
            if (author.discriminator && author.discriminator !== "0") {
              authorTag = `${author.username}#${author.discriminator}`;
            } else {
              authorTag = author.username ?? author.tag ?? "Inconnu";
            }
          }
          // Contenu du message (texte brut, peut être vide si embed/attachment only)
          content = msg.content ?? "";
          if (!content && msg.embeds?.length) {
            // Fallback : titre du premier embed si pas de contenu texte
            content = msg.embeds[0]?.title ?? msg.embeds[0]?.description ?? "";
          }
          content = content.slice(0, 500);
        }
      }
    } catch (err) {
      console.error("[MSGBM] Erreur fetch message :", err.message);
    }

    bookmarks.push({
      messageId,
      channelId,
      guildId:  guildId === "@me" ? null : guildId,
      authorTag,
      content,
      url:     payload.url,
      savedAt: Date.now(),
      note:    payload.note || null,
    });
    if (bookmarks.length > 200) bookmarks.shift();
    save(bookmarks);
    return { bookmarks };
  }

  if (action === "clear") {
    save([]);
    return { bookmarks: [] };
  }

  if (action === "remove") {
    const idx = (index ?? 1) - 1;
    if (idx < 0 || idx >= bookmarks.length)
      throw new Error(`Index invalide (1–${bookmarks.length}).`);
    bookmarks.splice(idx, 1);
    save(bookmarks);
    return { bookmarks };
  }

  if (action === "note") {
    const idx = (index ?? 1) - 1;
    if (idx < 0 || idx >= bookmarks.length)
      throw new Error(`Index invalide (1–${bookmarks.length}).`);
    bookmarks[idx].note = note || null;
    save(bookmarks);
    return { bookmarks };
  }

  throw new Error(`Action msgbookmarks inconnue : '${action}'`);
}

// ── Listener messageCreate pour sauvegarder les messages bookmarkés ──────────

async function handleBookmarkReaction(message, client) {
  const bookmarks = load();

  const author = message.author;
  let authorTag = "Inconnu";
  if (author) {
    if (author.discriminator && author.discriminator !== "0") {
      authorTag = `${author.username}#${author.discriminator}`;
    } else {
      authorTag = author.username ?? author.tag ?? "Inconnu";
    }
  }

  bookmarks.push({
    messageId:  message.id,
    channelId:  message.channel.id,
    guildId:    message.guild?.id ?? null,
    authorTag,
    content:    (message.content ?? "").slice(0, 500),
    url:        message.url,
    savedAt:    Date.now(),
    note:       null,
  });
  if (bookmarks.length > 200) bookmarks.shift();
  save(bookmarks);
}

module.exports = { name: "msgbookmarks", execute, handleBookmarkReaction };

File: packages/sb-uhq/app/commands/utilitaires/msgbookmarks.py
"""msgbookmarks — messages importants sauvegardés avec notes."""

from __future__ import annotations

import re
import time

from ...func.data_path import data_path, read_json, write_json
from ...func.discord_util import fetch_channel, user_tag
from ...func.logbus import logerr

MSGBM_FILE = data_path("config", "msgbookmarks.json")

# Ancré sur l'URL complète : un `search` non ancré acceptait
# https://exemple.test/discord.com/channels/1/2/3, et cette URL arbitraire était
# ensuite stockée puis affichée comme lien cliquable dans le panel.
_URL_RE = re.compile(
    r"^https://(?:(?:canary|ptb)\.)?discord(?:app)?\.com/channels/(\d{1,20}|@me)/(\d{1,20})/(\d{1,20})/?$")


def _load() -> list:
    return read_json(MSGBM_FILE, [])


def _save(data) -> None:
    write_json(MSGBM_FILE, data)


def _now_ms() -> int:
    return int(time.time() * 1000)


async def execute(client, payload):
    action = payload.get("action")
    index = payload.get("index")
    note = payload.get("note")
    bookmarks = _load()

    if action == "list":
        return {"bookmarks": bookmarks}

    if action == "add":
        url = payload.get("url")
        if not url:
            raise ValueError("url requis.")
        match = _URL_RE.match(str(url).strip())
        if not match:
            raise ValueError("URL de message Discord invalide.")
        guild_id, channel_id, message_id = match.group(1), match.group(2), match.group(3)

        author_tag = "Inconnu"
        content = ""
        try:
            channel = await fetch_channel(client, channel_id)
            if channel:
                msg = await channel.fetch_message(int(message_id))
                if msg:
                    author_tag = user_tag(msg.author) or "Inconnu"
                    content = msg.content or ""
                    if not content and msg.embeds:
                        e = msg.embeds[0]
                        content = getattr(e, "title", None) or getattr(e, "description", None) or ""
                    content = content[:500]
        except Exception as err:  # noqa: BLE001
            logerr(f"[MSGBM] Erreur fetch message : {err}")

        bookmarks.append({
            "messageId": message_id,
            "channelId": channel_id,
            "guildId": None if guild_id == "@me" else guild_id,
            "authorTag": author_tag,
            "content": content,
            "url": url,
            "savedAt": _now_ms(),
            "note": payload.get("note") or None,
        })
        if len(bookmarks) > 200:
            bookmarks.pop(0)
        _save(bookmarks)
        return {"bookmarks": bookmarks}

    if action == "clear":
        _save([])
        return {"bookmarks": []}

    if action == "remove":
        idx = (index or 1) - 1
        if idx < 0 or idx >= len(bookmarks):
            raise ValueError(f"Index invalide (1–{len(bookmarks)}).")
        bookmarks.pop(idx)
        _save(bookmarks)
        return {"bookmarks": bookmarks}

    if action == "note":
        idx = (index or 1) - 1
        if idx < 0 or idx >= len(bookmarks):
            raise ValueError(f"Index invalide (1–{len(bookmarks)}).")
        bookmarks[idx]["note"] = note or None
        _save(bookmarks)
        return {"bookmarks": bookmarks}

    raise ValueError(f"Action msgbookmarks inconnue : '{action}'")
